subprojects/model-core/src/main/java/org/gradle/internal/instantiation/AsmBackedClassGenerator.java

```diff
@@ -1049,6 +1086,26 @@ public void applyReadOnlyManagedStateToGetter(PropertyMetadata property, Method
             });
         }
 
+        @Override
+        public void applyNestedManagedStateToGetter(PropertyMetadata property, Method getter) {
+            // GENERATE public <type> <getter>() { if (<field> == null) { <field> = getInstantiator().newInstance(type); } return <field> }
```

Comment:
Why the lazy init? Why can't we generate these up-front instead? That would remove the need to pass the injector, no?

Reply:
We can initialize the nested fields in the constructor. That would cause a nested call to the instantiator when the outer class is instantiated. I can try that. The strange thing is then that the fields with property types (`Property`, `FileCollection`, etc.) are initialized lazily while the nested beans are not. I would either do both in the constructor or none. I went for the lazy initialization since the other managed properties are also initialized lazily.

Reply:
I talked to Adam, and we want to have the instantiation as cheap as possible. Services should be lazy for sure. The reason is that configuration time is serial while querying the properties happens at execution time in parallel.